packages/opencensus-web-instrumentation-zone/src/interaction-tracker.ts

```diff
@@ -90,43 +96,140 @@ export class InteractionTracker {
         return runTask.call(taskZone as {}, task, applyThis, applyArgs);
       } finally {
         console.log('Run task finished.');
-        console.log('Time to complete: ' + (Date.now() - time));
+        if (
+          interceptingElement ||
+          (shouldCountTask(task) && isTrackedTask(task))
+        ) {
+          this.decrementTaskCount(task.zone.get('traceId'));
+        }
       }
     };
 
     const scheduleTask = Zone.prototype.scheduleTask;
-    Zone.prototype.scheduleTask = function<T extends Task>(task: T): T {
+    Zone.prototype.scheduleTask = <T extends Task>(task: T) => {
       console.warn('Scheduling task');
       console.log(task);
 
-      let taskZone: Zone = this;
-      if (task.zone && task.zone && task.zone.get('isTracingZone')) {
+      let taskZone = Zone.current;
+      if (isTrackedTask(task)) {
         taskZone = task.zone;
       }
       try {
         return scheduleTask.call(taskZone as {}, task) as T;
       } finally {
+        if (shouldCountTask(task) && isTrackedTask(task)) {
+          this.incrementTaskCount(task.zone.get('traceId'));
+        }
+        console.warn('Finished Scheduling task');
       }
     };
 
     const cancelTask = Zone.prototype.cancelTask;
-    Zone.prototype.cancelTask = function(task: AsyncTask) {
+    Zone.prototype.cancelTask = (task: AsyncTask) => {
       console.warn('Cancel task');
       console.log(task);
 
-      let taskZone: Zone = this;
-      if (task.zone && task.zone.get('isTracingZone')) {
+      let taskZone = Zone.current;
+      if (isTrackedTask(task)) {
         taskZone = task.zone;
       }
 
       try {
         return cancelTask.call(taskZone as {}, task);
       } finally {
+        if (isTrackedTask(task) && shouldCountTask(task)) {
+          this.decrementTaskCount(task.zone.get('traceId'));
+        }
+        console.warn('Finished cancel task');
       }
     };
   }
+
+  /** Increments the count of outstanding tasks for a given interaction id. */
+  private incrementTaskCount(interactionId: string) {
+    const stopWatch = this.getStopwatch(interactionId);
+    if (!stopWatch) return;
+    stopWatch.incrementTaskCount();
+
+    if (interactionId in this.interactionCompletionTimeouts) {
+      // This may get called a lot. Run in the root zone.
+      Zone.root.run(() => {
+        clearTimeout(this.interactionCompletionTimeouts[interactionId]);
```

Comment:
What's the role of this interaction completion timeout? Why isn't it sufficient to just count tasks as they are scheduled/completed/cancelled?

Reply:
We need this to instead of declaring an interaction to be complete when the number of active interactions reaches 0 we add a task to the queue that will actually complete the interaction if none of the tasks scheduled ahead of it try and increment the task counter for the given interaction id.

Sometimes the count of tasks might be 0 but the interaction has more tasks.

Reply:
OK, could you add a comment to explain all that?